src/dividat-driver/firmware/main.go

```diff
@@ -100,35 +108,40 @@ func Update(parentCtx context.Context, image io.Reader, deviceSerial *string, co
 			// We do need to rediscover, as the legacy device may still just have
 			// restarted into the bootloader and obtained a new IP address.
 			ctx, cancel := context.WithTimeout(parentCtx, 60*time.Second)
-			legacyDiscoveredAddr, err := discover("_sensoControl._tcp", deviceSerial, ctx)
+			legacyDiscoveredAddr, err := discover("_sensoControl._tcp", deviceSerial, ctx, onProgress)
 			cancel()
 			if err == nil {
 				dfuHost = legacyDiscoveredAddr
+				onProgress("Senso discovered via _sensoControl._tcp")
 			} else if controllerHost != "" {
-				fmt.Printf("Could not discover update service, trying to fall back to previous discovery %s.\n", controllerHost)
+				onProgress(fmt.Sprintf("Could not discover update service, trying to fall back to previous discovery %s.", controllerHost))
 				dfuHost = controllerHost
 			} else {
-				fail = fmt.Errorf("Could not find any Senso bootloader to transmit firmware to: %s", err)
+				msg := fmt.Sprintf("Could not find any Senso bootloader to transmit firmware to: %s", err)
+				onProgress(msg)
+				fail = fmt.Errorf(msg)
 				return
 			}
 		} else {
 			dfuHost = discoveredAddr
+			onProgress("Senso discovered via _sensoUpdate._udp")
 		}
 	}
 
+	onProgress("Preparing to transmit firmware.")
 	// 4: Transmit firmware via TFTP
 	time.Sleep(5 * time.Second) // Wait to ensure proper TFTP startup
```

Comment:
This 5 seconds sleep might be a bit short if we are working with a `configuredAddr`. We get a TFTP timeout that looks like this:

![image](https://github.com/dividat/driver/assets/47458/a1add3e3-438d-4d7f-810a-4d470db52f5f)

Using 15 seconds worked for me to have the procedure work every time with a fixed address. Otherwise it only worked reliably when discovering the service instead of specifying a fixed address.

Reply:
I increased the timeout in [Wait 15 seconds for tftp startup](https://github.com/dividat/driver/pull/127/commits/4201904cb5da062b37efd052668b3d871370b80c). However, I ran multiple stress tests with the latest changes and with only five seconds, and it seems to be enough, so I reduced the timeout back down to 5s. I wonder if this will prove to be robust enough under different network conditions though, so more testing is required.